Feature/tw automatic group order invoice generation

Viehlieb opened the pull request:
Offers the possibility to generate Invoices for given Group Orders.
Generation is possible after an order is closed.
If order is 'settled' then there's additionally the possibility to let the software send e-mails automatically on button press. The corresponding invoices will be sent as attachment via mail to the order group members.

Notes:
To use automatic sending of invoices, select option in "Admin/Finance" config
You can enter a "payment method " in "Admin/Finance"
To make things work, it is mandatory to enter a tax_number in "Admin/Foodcoop" config

As for now "payment method" determines the "body" of the invoice. 
An example of new options and the resulting invoice can be found here:

![tax_number](https://user-images.githubusercontent.com/20905449/147212544-d483a45e-11fd-4dec-a7f8-ce443f1bf7ca.png)
![finance](https://user-images.githubusercontent.com/20905449/147212550-24e855e8-4af1-48fd-8284-e170dec669e1.png)
![invoice](https://user-images.githubusercontent.com/20905449/147212553-d6b97b04-12e9-4ba1-ab33-9dd8b505a9d8.png)

Additionally for older and already settled orders, group order invoices can be generated manually under "finance/balancing"
![generate](https://user-images.githubusercontent.com/20905449/147212881-0e5d7e0a-7e72-4889-a924-c2cf216d3927.png)
![download](https://user-images.githubusercontent.com/20905449/147212884-3d7ccb58-3bbe-4d7e-afec-1b75ad591bfd.png)

Invoices will be generated given the momentarily given information about invoicee and invoicer (tax_number, contact details...). SO take care when delting an invoice.

wvengen commented:
This seems like a great feature. Thanks for taking the time to draft this!
What exactly is the use of removing an invoice?

Viehlieb commented:
> 
Hey. Excuse my late reply.
When incorrect settings are set in the admin configs, there should be the possibility to delete the old and generate a new invoice with correct settings applied.

Viehlieb reviewed (commented):
If I provide a link to article details it was not possible to click the link because of the :hover feature. 
Behavior before commit:
Hover over articles would render the article infos and if no article is hovered there's a green box showing
Behavior after commit:
Hover over articles will render article infos asbefore.
Click article, and hover over other articleswill render article infos as before.
BUT the 'clicked' info will be the 'fallback info' which will remain in the green box (article clicked last). So it is possible to click alink in the green bo
![Screenshot from 2022-04-05 13-24-25](https://user-images.githubusercontent.com/20905449/161743764-14020bdb-4f69-40cb-8233-a18bbe2f1cbf.png)
x

Viehlieb commented:
Is there anything more needed to merge this feature into master branch?

paroga commented:
first: sorry for the late reply
second: thanks for the contribution
third: this is a huge change and therefore hard to review: could split it smaller self-containing parts? e.g.: change the model in a separate commit

there are a bunch of style violations and i have no idea why they are not marked by the CI system :disappointed:
if i'm not wrong there are a bunch of changes which require cleanup before merge. it seams there a some dummy "hello world" files and debug outputs introduces.
please feel free to correct my if i'm wrong. i just skimmed the PR

Viehlieb commented:
Hey. Thanks for the hints. 
It made me look into the code again and I am working on fixing the points you mentioned.

1) I noticed, there was no authentication clause to deal wirth group order invoices.
With b6854bc41609fe42bcb5bae733f9ffc4524d3ea6 I added a neccessary update of the group_order_invoice_controller.
:authenticate_finance, so only for finance authenticated users can see/download and generate group_order_invoices.
2) Removed the "Hello World" Dummy files.
3) Todo Style issues
Let me know, if there is anything more you notice.

paroga commented on app/controllers/group_order_invoices_controller.rb:
```diff
@@ -0,0 +1,38 @@
+        end
+      end
+    else raise RecordInvalid
+      redirect_back fallback_location: root_path, notice: 'Something went wrong', :alert => I18n.t('errors.general_msg', :msg => "#{error} " + I18n.t('errors.check_tax_number'))
```
* what kind of error do you want to catch here?
* please use translations for user visible strings.
* please do not use `notice` and `alert` at the same time
* there is usually no need to use `I18n.t` and you could use just `t` in the controllers
* please use newer syntax `name: value` instead of `:name => value`

paroga commented on app/controllers/group_order_invoices_controller.rb:
```diff
@@ -0,0 +1,38 @@
+    if FoodsoftConfig[:contact][:tax_number]
+      respond_to do |format|
+        format.pdf do
+          send_group_order_invoice_pdf @group_order_invoice if FoodsoftConfig[:contact][:tax_number]
```
you check already for the `tax_number` in line 6. what's the purpose of the second check here?

paroga commented on app/controllers/group_order_invoices_controller.rb:
```diff
@@ -0,0 +1,38 @@
+
+  def show
+    @group_order_invoice = GroupOrderInvoice.find(params[:id])
+    if FoodsoftConfig[:contact][:tax_number]
```
why can't we show an invoice without an tax number?

paroga commented on app/controllers/group_order_invoices_controller.rb:
```diff
@@ -0,0 +1,38 @@
+
+  def destroy
+    goi = GroupOrderInvoice.find(params[:id])
+    @order = goi.group_order.order
```
what's the purpose of `@order` here?

paroga commented on app/controllers/group_order_invoices_controller.rb:
```diff
@@ -0,0 +1,38 @@
+
+  def create
+    go = GroupOrder.find(params[:group_order])
+    @order = go.order
```
what's the purpose of @order here?

paroga commented on app/controllers/group_order_invoices_controller.rb:
```diff
@@ -0,0 +1,38 @@
+  def create
+    go = GroupOrder.find(params[:group_order])
+    @order = go.order
+    goi = GroupOrderInvoice.find_or_create_by!(group_order_id: go.id)
```
why `find_or_create` in the `create` method?

paroga commented on app/controllers/group_order_invoices_controller.rb:
```diff
@@ -0,0 +1,38 @@
+    respond_to do |format|
+      format.js
+    end
+    redirect_back fallback_location: root_path
```
why do you redirect here to `root_path`?

paroga commented on app/controllers/group_order_invoices_controller.rb:
```diff
@@ -0,0 +1,38 @@
+    end
+    redirect_back fallback_location: root_path
+  rescue => error
+    redirect_back fallback_location: root_path, notice: 'Something went wrong', :alert => I18n.t('errors.general_msg', :msg => error)
```
the comments form line 13 apply here too

paroga commented on app/models/group_order_invoice.rb:
```diff
@@ -0,0 +1,60 @@
+  def generate_invoice_number(count)
+    trailing_number = count.to_s.rjust(4, '0')
+    unless  GroupOrderInvoice.find_by(invoice_number: Time.now.strftime("%Y%m%d") + trailing_number)
+      Time.now.strftime("%Y%m%d") + trailing_number
```
* is there a way to make the format configureable?
* please create a variable for values you use multiple times

paroga commented on app/models/group_order_invoice.rb:
```diff
@@ -0,0 +1,60 @@
+    unless  GroupOrderInvoice.find_by(invoice_number: Time.now.strftime("%Y%m%d") + trailing_number)
+      Time.now.strftime("%Y%m%d") + trailing_number
+    else
+      generate_invoice_number(count.to_i + 1)
```
* recursive calling `generate_invoice_number` here is a very bad idea, since you create a database query all the time which is slow and blocks everything else
* you could either store the invoice number somewhere in the database or find the last invoice via a single query to the database (something like `SELECT invoice_number ... WHER invoice_number LIKE '...%' ORDER BY invoice_number DESC LIMIT 1`)

paroga commented on app/models/group_order_invoice.rb:
```diff
@@ -0,0 +1,60 @@
+  end
+
+  def tax_number_set
+    if FoodsoftConfig[:contact][:tax_number].blank?
+      errors.add(:group_order_invoice, "Keine Steuernummer in FoodsoftConfig :contact gesetzt")
+    end
```
this method does not access the database and therefore should be removed from the model and move to a better location (BTW: there are user visible string which are not translated)

paroga commented on app/models/group_order_invoice.rb:
```diff
@@ -0,0 +1,60 @@
+  def init
+    self.invoice_number = generate_invoice_number(1) unless self.invoice_number
+    self.invoice_date = Time.now unless self.invoice_date
+    self.payment_method = FoodsoftConfig[:group_order_invoices]&.[](:payment_method) || I18n.t('activerecord.attributes.group_order_invoice.payment_method') unless self.payment_method
```
why is the payment method stored in `FoodsoftConfig`?
please do not use translated strings in the model. depending on the locale of the current user different data would be used and stored in the database

paroga commented on app/models/group_order_invoice.rb:
```diff
@@ -0,0 +1,60 @@
+  end
+
+  def init
+    self.invoice_number = generate_invoice_number(1) unless self.invoice_number
```
please use a `before_validation` hook instead (see e.g. [`set_password` in the `User` model](https://github.com/foodcoops/foodsoft/blob/f7f8774c433d58bcdf0c03fe93a7967fcf0e6437/app/models/user.rb#L122-L127))

paroga commented on app/models/group_order_invoice.rb:
```diff
@@ -0,0 +1,60 @@
+  end
+
+  def name
+    I18n.t('activerecord.attributes.group_order_invoice.name') + "_#{invoice_number}"
```
the translation should be moved from the model to a controller, since the model should be language independent

paroga commented on app/models/group_order_invoice.rb:
```diff
@@ -0,0 +1,60 @@
+    I18n.t('activerecord.attributes.group_order_invoice.name') + "_#{invoice_number}"
+  end
+
+  def load_data_for_invoice
```
why do you copy everything here? usually working with a `GroupOrderInvoice` instance should be the way to go.

paroga reviewed (changes requested):
i looked only into the first commit of your PR for now

RayOei commented:
As far as I can make out, no follow-up exists. 
I do not know the actual use case but I am inclined to close this in its current state?
_Added tags for review._
